packages/terra-disclosure-manager/src/DisclosureManager.jsx

```diff
@@ -0,0 +1,351 @@
+import React from 'react';
+import PropTypes from 'prop-types';
+import AppDelegate from 'terra-app-delegate';
+
+const availableDisclosureSizes = {
+  TINY: 'tiny',
+  SMALL: 'small',
+  MEDIUM: 'medium',
+  LARGE: 'large',
+  HUGE: 'huge',
+  FULLSCREEN: 'fullscreen',
+};
+
+export { availableDisclosureSizes };
+
+const propTypes = {
+  /**
+   * An AppDelegate instance that will be integrated with the DisclosureManager instance. The DisclosureManager will defer to it if unsupported
+   * actions occur.
+   */
+  app: AppDelegate.propType,
+  /**
+   * The child components that will be provided with an AppDelegate 'app' prop used to interact with the DisclosureManager instance.
+   */
+  children: PropTypes.node,
+  /**
+   * A function used to provide rendering capabilities to the DisclosureManager.
+   */
+  render: PropTypes.func.isRequired,
+  /**
+   * An array of disclosure types that the DisclosureManager should support. If an unsupported disclosure request occurs, the DisclosureManager will
+   * utilize its 'app' prop and forward the request instead of handling the request itself.
+   */
+  supportedDisclosureTypes: PropTypes.array,
+};
+
+const defaultProps = {
+  supportedDisclosureTypes: [],
+};
+
+class DisclosureManager extends React.Component {
+  /**
+   * Clones the current disclosure component state objects and returns the structure for further mutation.
+   */
+  static cloneDisclosureState(state) {
+    const newState = Object.assign({}, state);
+    newState.disclosureComponentKeys = Object.assign([], newState.disclosureComponentKeys);
+    newState.disclosureComponentData = Object.assign({}, newState.disclosureComponentData);
+
+    return newState;
+  }
+
+  constructor(props) {
+    super(props);
+
+    this.resolveDismissPromise = this.resolveDismissPromise.bind(this);
+    this.resolveDismissChecksInSequence = this.resolveDismissChecksInSequence.bind(this);
+
+    this.safelyCloseDisclosure = this.safelyCloseDisclosure.bind(this);
+    this.generatePopFunction = this.generatePopFunction.bind(this);
+
+    this.renderContentComponents = this.renderContentComponents.bind(this);
+    this.renderDisclosureComponents = this.renderDisclosureComponents.bind(this);
+
+    this.openDisclosure = this.openDisclosure.bind(this);
+    this.pushDisclosure = this.pushDisclosure.bind(this);
+    this.popDisclosure = this.popDisclosure.bind(this);
+    this.closeDisclosure = this.closeDisclosure.bind(this);
+    this.requestDisclosureFocus = this.requestDisclosureFocus.bind(this);
+    this.releaseDisclosureFocus = this.releaseDisclosureFocus.bind(this);
+    this.maximizeDisclosure = this.maximizeDisclosure.bind(this);
+    this.minimizeDisclosure = this.minimizeDisclosure.bind(this);
+
+    // These cached functions are stored outside of state to prevent unnecessary rerenders.
+    this.dismissChecks = {};
```

Comment:
Do you have any concern with these becoming unaligned with the state?

Reply:
The registration of one of these functions shouldn't be a renderable event, which is why they're not stored in state. However, this object is cleaned up with the other state changing events to ensure state functions aren't hanging around.